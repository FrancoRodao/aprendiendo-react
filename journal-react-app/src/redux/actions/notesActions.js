import { db } from "../../firebase/firebase.config"
import { imageUploadToCloudinary } from "../../helpers/imageUpload"
import { loadNotes } from "../../helpers/loadNotes"
import { types } from "../types/types"

export const startAddNewNote = () => {
    return async (dispatch, getState) => {
        const userId = getState().auth.uid

        const newNote = {
            title: 'title',
            body: 'body',
            date: new Date().getTime()
        }

        const noteRef = await db.collection(`${userId}/journal/notes`).add(newNote)
        dispatch(addNewNote(noteRef.id, newNote))
        dispatch(activeNoteAction(noteRef.id, newNote))
    }
}

export const addNewNote = (noteId, note) => ({
    type: types.notesAdd,
    payload: {
        id: noteId,
        note
    }
})

export const activeNoteAction = (id, note) => ({
    type: types.notesSetActive,
    payload: {
        id,
        ...note
    }
})

export const startLoadingNotes = (userId) => {
    return async (dispatch) => {
        const notes = await loadNotes(userId)
        dispatch(setNotes(notes))
    }
}

export const setNotes = (notes) => ({
    type: types.notesLoad,
    payload: notes
})


export const startSaveNote = (note) => {
    return async (dispatch, getState) => {
        const { uid } = getState().auth
        const noteToFireStore = {...note}
        delete noteToFireStore.id

        await db.collection(`${uid}/journal/notes`).doc(`${note.id}`).update(noteToFireStore)
        dispatch(refreshNote(note))
    }
}

export const refreshNote = (note) => ({
    type: types.notesUpdate,
    payload: {
        note
    }
})

export const startUploadPicture = (file) => {
    return async (dispatch, getState) => {
        const activeNote = getState().notes.active
        const imageUrl = await imageUploadToCloudinary(file)
        activeNote.imgUrl = imageUrl
        
        dispatch(startSaveNote(activeNote))
    }
}


export const startDeleteNote = (note) => {
    return async (dispatch, getState) => {
        const userId = getState().auth.uid

        await db.doc(`${userId}/journal/notes/${note.id}`).delete()
        dispatch(deleteNote(note))
    }
}

export const deleteNote = (note) => ({
    type: types.notesDelete,
    payload: {
        note
    }
})