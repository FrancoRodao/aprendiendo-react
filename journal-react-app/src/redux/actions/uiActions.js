import { types } from "../types/types"

export const startLoadingAction = () => (
    {
        type: types.uiStartLoading,
        payload: {
            loading: true
        }
    }
)


export const finishLoadingAction = () => (
    {
        type: types.uiFinishLoading,
        payload: {
            loading: false
        }
    }
)

export const errorAction = (state, message) => (
    {
        type: types.uiError,
        payload: {
            state,
            lastMsg: message
        }
    }
)

export const clearErrorsAction = () =>  (
    {
        type: types.uiError,
        payload: {
            state: false,
            lastMsg: ""
        }
    }
)