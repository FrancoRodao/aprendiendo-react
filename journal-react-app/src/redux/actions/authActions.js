import { types } from "../types/types"
import { firebase, googleAuthProvider } from "../../firebase/firebase.config";
import { clearErrorsAction, errorAction, finishLoadingAction, startLoadingAction } from "./uiActions";

export const startLogin = (email, password) => {
    return (dispatch) => {
        dispatch(startLoadingAction())

        return firebase.auth().signInWithEmailAndPassword(email, password)
            .then(({ user }) => {
                dispatch(loginAction(user.uid, user.displayName))
                dispatch(clearErrorsAction())
            })
            .catch(e => {
                if (e.code === "auth/user-not-found" || "auth/wrong-password") {
                    dispatch(errorAction(true, "Wrong credentials"))
                    return
                }
                dispatch(errorAction(true, "Unexpected error"))
            })
            .finally(() => dispatch(finishLoadingAction()))
    }
}

export const startGoogleLogin = () => {
    return dispatch => {
        dispatch(startLoadingAction())

        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({ user }) => {
                dispatch(loginAction(user.uid, user.displayName))
                dispatch(clearErrorsAction())
            })
            .catch(() => dispatch(errorAction(true, "Unexpected error")))
            .finally(() => dispatch(finishLoadingAction()))
    }
}

export const loginAction = (uid, displayname) => (
    {
        type: types.authLogin,
        payload: {
            uid,
            displayname
        }
    }
)

export const startRegisterUser = (name, email, password) => {
    return (dispatch, state) => {
        dispatch(startLoadingAction())
        return firebase.auth().createUserWithEmailAndPassword(email, password)
            .then(async ({ user }) => {
                await user.updateProfile({
                    displayName: name,
                    photoURL: ''
                })
                dispatch(registerUserAction(user.uid, user.displayName))
                dispatch( clearErrorsAction() )
            })
            .catch(e => {
                if (e.code === "auth/email-already-in-use") {
                    dispatch(errorAction(true, "Email already registered"))
                    return
                }
                dispatch(errorAction(true, "Unexpected error"))
            })
            .then(() => state())
            .finally(() => dispatch(finishLoadingAction()))
        
    }
}

export const registerUserAction = (uid, displayname) => {
    return {
        type: types.register,
        payload: {
            uid,
            displayname
        }
    }
}


export const startLogout = () => {
    return async (dispatch) => {
        dispatch(startLoadingAction())

        firebase.auth().signOut()
            .then(() => dispatch(logoutUserAction()))
            .catch(() => dispatch(errorAction(true, "Unexpected error")))
            .finally(() => dispatch(finishLoadingAction()))
    }
}

export const logoutUserAction = () => ({
    type: types.authLogout,
    payload: {}
})

