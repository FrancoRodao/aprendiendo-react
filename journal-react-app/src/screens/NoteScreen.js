import React from 'react'
import { useRef } from 'react'
import { useEffect } from 'react'
import { useDispatch } from 'react-redux'
import { useSelector } from 'react-redux'
import NotesAppBar from '../components/notes/NotesAppBar'
import useForm from '../hooks/useForm'
import { activeNoteAction, startDeleteNote } from '../redux/actions/notesActions'

export default function NoteScreen() {

    const dispatch = useDispatch()

    const { active: activeNote } = useSelector(state => state.notes)
    const activeNoteId = useRef(activeNote.id)

    const {formValues, handleInputChange, resetForm} = useForm(activeNote)

    const handleDelete = () => {
        dispatch(startDeleteNote(activeNote))
    }
    
    //reset form when change active note
    useEffect(() => {
        if(activeNoteId.current !== activeNote.id){
            resetForm()
            activeNoteId.current = activeNote.id
        }
    }, [activeNote, resetForm])

    //update values
    useEffect(() => {
        dispatch(activeNoteAction(formValues.id, {...formValues}))
    }, [formValues, dispatch])

    return (
        <div className="notes__main-content">

            <NotesAppBar />

            <div className="notes__content">

                <input
                    type="text"
                    placeholder="Some awesome title"
                    className="notes__title-input"
                    autoComplete="off"
                    name="title"
                    onChange={handleInputChange}
                    value={formValues.title}
                />

                <textarea
                    name="body"
                    onChange={handleInputChange}
                    value={formValues.body}
                    placeholder="What happend today"
                    className="notes__textarea"
                    autoComplete="off"
                />

                {
                    activeNote.imgUrl &&
                    <div className="notes__image">
                        <img
                            src={activeNote.imgUrl}
                            alt="note"
                        />
                    </div>
                }

            </div>

            <button onClick={handleDelete} className="btn btn-danger">
                Delete
            </button>

        </div>
    )
}
