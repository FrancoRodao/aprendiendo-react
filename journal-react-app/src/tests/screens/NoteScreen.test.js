import { mount } from "enzyme"

import { Provider } from "react-redux"
import thunk from "redux-thunk"
import createMockStore from "redux-mock-store"

import NoteScreen from "../../screens/NoteScreen"

import { activeNoteAction } from "../../redux/actions/notesActions"
jest.mock('../../redux/actions/notesActions', () => ({
    activeNoteAction: jest.fn()
}))

const storeInitialState = {
    auth: {},
    notes: {
        active: {
            id: 7357,
            title: 'hi',
            body: 'world',
            imgUrl: 'https://someurl.com'
        }
    },
    ui: {}
}

const middlewares = [thunk]
const mockStore = createMockStore(middlewares)
const store = mockStore(storeInitialState)
store.dispatch = jest.fn()

describe('NoteScreen tests', () => {
    const wrapper = mount(
        <Provider store={store}>
            <NoteScreen />
        </Provider>)
    test('should to match snapshot', () => {
        expect(wrapper).toMatchSnapshot()
    })

    test('should update active note', async () => {

        wrapper.find('input[name="title"]').simulate('change', {
            target: {
                name: 'title',
                value: 'new value',
            }
        })

        expect(activeNoteAction).toBeCalledWith(7357, {
            id: 7357,
            title: 'new value',
            body: 'world',
            imgUrl: 'https://someurl.com'
        })

    })
})
