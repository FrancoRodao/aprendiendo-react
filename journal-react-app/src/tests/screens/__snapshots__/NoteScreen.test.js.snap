// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteScreen tests should to match snapshot 1`] = `
<div
  className="notes__main-content"
>
  <div
    className="notes__appbar"
  >
    <span>
      28 de agosto 2020
    </span>
    <input
      id="fileSelector"
      name="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some awesome title"
      type="text"
      value="hi"
    />
    <textarea
      autoComplete="off"
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="What happend today"
      value="world"
    />
    <div
      className="notes__image"
    >
      <img
        alt="note"
        src="https://someurl.com"
      />
    </div>
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
